astropy/stats/funcs.py

```diff
@@ -761,22 +773,51 @@ def median_absolute_deviation(a, axis=None):
     # See https://github.com/numpy/numpy/issues/7330 why using np.ma.median
     # for normal arrays should not be done (summary: np.ma.median always
     # returns an masked array even if the result should be scalar). (#4658)
-    if isinstance(a, np.ma.MaskedArray):
+    if isinstance(data, np.ma.MaskedArray):
         func = np.ma.median
+        if ignore_nan:
+            data = np.ma.masked_where(np.isnan(data), data)
+        elif (NUMPY_LT_1_10 or not NUMPY_LT_1_12) and np.any(np.isnan(data)):
+            # masked arrays containing NaNs should return NaN, but they only
+            # do in 1.11 because of a bug:
+            # https://github.com/astropy/astropy/pull/5232#issuecomment-264751887
+            # TODO for devs: numpy 1.13 may fix this, in which case the
+            # special-casing can be removed
+            return np.nan
+    elif ignore_nan:
+        if NUMPY_LT_1_10 or axis is not None:
+            # older versions don't have nanmedian, but nanmedian
+            # may be faster than np.ma.median *if* axis is None
+            # https://gist.github.com/MSeifert04/1551e80e1aa85b48c9191e727c38f9a9
+            data = np.ma.masked_where(np.isnan(data), data)
+            func = np.ma.median
+        else:
+            # faster when axis is None & numpy is recent
+            func = np.nanmedian
     else:
         func = np.median
 
-    a = np.asanyarray(a)
-    a_median = func(a, axis=axis)
+    if not ignore_nan and NUMPY_LT_1_10 and np.any(np.isnan(data)):
+        warn("Numpy versions <1.10 will return a number rather than NaN for "
+             "the median of arrays containing NaNs.  This behavior is "
+             "unlikely to be what you expect.")
+
+    data = np.asanyarray(data)
+    data_median = func(data, axis=axis)
 
     # broadcast the median array before subtraction
     if axis is not None:
-        a_median = np.expand_dims(a_median, axis=axis)
+        data_median = np.expand_dims(data_median, axis=axis)
+
+    result = func(np.abs(data - data_median), axis=axis)
```

Comment:
as you are working on a copy here you can use `overwrite_input=True` here to safe some memory and time.
Additionally the abs could work inplace on the subtracted result to safe more memory.